File: src/nanogridbot/core/container_runner.py
"""Docker container runner for executing Claude Agent."""

import asyncio
import json
import time
from pathlib import Path
from typing import Any, Literal

from nanogridbot.config import get_config
from nanogridbot.core.mount_security import validate_group_mounts
from nanogridbot.types import ContainerConfig, ContainerOutput
from nanogridbot.utils.formatting import format_messages_xml

# Output markers for parsing
OUTPUT_START_MARKER = "---NANOGRIDBOT_OUTPUT_START---"
OUTPUT_END_MARKER = "---NANOGRIDBOT_OUTPUT_END---"


async def run_container_agent(
    group_folder: str,
    prompt: str,
    session_id: str | None,
    chat_jid: str,
    is_main: bool = False,
    container_config: ContainerConfig | None = None,
    timeout: int | None = None,
    env: dict[str, str] | None = None,
) -> ContainerOutput:
    """Run Claude Agent in a Docker container.

    Args:
        group_folder: Group folder name
        prompt: Prompt to send to the agent
        session_id: Existing session ID if any
        chat_jid: Chat JID for context
        is_main: Whether this is the main group
        container_config: Optional container configuration
        timeout: Optional timeout in seconds
        env: Optional environment variables for container

    Returns:
        ContainerOutput with execution result
    """
    from loguru import logger

    config = get_config()

    # Merge environment variables from container_config and explicit env parameter
    merged_env: dict[str, str] = {}
    if container_config and container_config.env:
        merged_env.update(container_config.env)
    if env:
        merged_env.update(env)

    # Extract channel from chat_jid
    channel = chat_jid.split(":")[0] if ":" in chat_jid else "unknown"

    # Record container start for metrics
    metric_id = None
    start_time = time.time()
    try:
        from nanogridbot.database import metrics as metrics_db

        metric_id = await metrics_db.record_container_start(group_folder, channel)
    except Exception:
        # Metrics are optional, don't fail if they can't be recorded
        pass

    # Build mounts
    try:
        mounts = await validate_group_mounts(
            group_folder=group_folder,
            container_config=container_config.model_dump() if container_config else None,
            is_main=is_main,
        )
    except Exception as e:
        logger.error(f"Mount validation failed: {e}")
        # Record failure
        if metric_id:
            try:
                await metrics_db.record_container_end(metric_id, "error", error=str(e))
            except Exception:
                pass
        return ContainerOutput(status="error", error=str(e))

    # Prepare input data
    input_data = {
        "prompt": prompt,
        "sessionId": session_id,
        "groupFolder": group_folder,
        "chatJid": chat_jid,
        "isMain": is_main,
    }

    # Build docker command
    cmd = build_docker_command(
        mounts=mounts,
        input_data=input_data,
        timeout=timeout or config.container_timeout,
        env=merged_env,
    )

    logger.debug(f"Starting container for {group_folder}")

    try:
        result = await _execute_container(cmd, input_data)
        # Record container end for metrics
        duration = time.time() - start_time
        status = "success" if result.status == "success" else "error"
        if metric_id:
            try:
                await metrics_db.record_container_end(
                    metric_id,
                    status=status,
                    duration_seconds=duration,
                    error=result.error,
                )
            except Exception:
                pass
        return result
    except Exception as e:
        logger.error(f"Container error: {e}")
        # Record failure
        duration = time.time() - start_time
        if metric_id:
            try:
                await metrics_db.record_container_end(metric_id, "error", duration_seconds=duration, error=str(e))
            except Exception:
                pass
        return ContainerOutput(status="error", error=str(e))


async def _execute_container(
    cmd: list[str],
    input_data: dict[str, Any],
) -> ContainerOutput:
    """Execute docker container and capture output.

    Args:
        cmd: Docker command arguments
        input_data: Input data to send to container

    Returns:
        ContainerOutput with result
    """
    from loguru import logger

    try:
        process = await asyncio.create_subprocess_exec(
            *cmd,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )

        # Write input to stdin
        input_json = json.dumps(input_data)
        process.stdin.write(input_json.encode())
        await process.stdin.drain()
        process.stdin.close()

        # Read output
        stdout, stderr = await asyncio.wait_for(
            process.communicate(),
            timeout=get_config().container_timeout,
        )

        # Parse output
        if stdout:
            output_str = stdout.decode("utf-8", errors="replace")
            result = _parse_output(output_str)
            if result:
                return result

        # Check for stderr errors
        if stderr:
            error_str = stderr.decode("utf-8", errors="replace")
            if error_str:
                logger.error(f"Container stderr: {error_str}")

        return ContainerOutput(
            status="error",
            error="No output from container",
        )

    except asyncio.TimeoutError:
        # Try to kill the process
        try:
            process.kill()
        except Exception:
            pass
        return ContainerOutput(
            status="error",
            error="Container execution timed out",
        )
    except FileNotFoundError:
        return ContainerOutput(
            status="error",
            error="Docker not found. Please install Docker.",
        )
    except Exception as e:
        return ContainerOutput(
            status="error",
            error=str(e),
        )


def _parse_output(output: str) -> ContainerOutput | None:
    """Parse container output.

    Args:
        output: Raw output string

    Returns:
        ContainerOutput or None if parsing fails
    """
    import json

    in_output = False
    output_lines = []

    for line in output.split("\n"):
        if OUTPUT_START_MARKER in line:
            in_output = True
            continue
        elif OUTPUT_END_MARKER in line:
            in_output = False
            break

        if in_output:
            output_lines.append(line)

    if output_lines:
        try:
            output_json = "\n".join(output_lines)
            data = json.loads(output_json)
            return ContainerOutput(
                status=data.get("status", "success"),
                result=data.get("result"),
                error=data.get("error"),
                new_session_id=data.get("newSessionId"),
            )
        except json.JSONDecodeError:
            # If not JSON, treat as result text
            return ContainerOutput(
                status="success",
                result="\n".join(output_lines),
            )

    return None


def build_docker_command(
    mounts: list[tuple[str, str, str]],
    input_data: dict[str, Any],
    timeout: int,
    env: dict[str, str] | None = None,
) -> list[str]:
    """Build docker run command.

    Args:
        mounts: List of (host_path, container_path, mode) tuples
        input_data: Input data to pass to container
        timeout: Timeout in seconds
        env: Optional environment variables for container

    Returns:
        Command as list of strings
    """
    from nanogridbot.config import get_config

    config = get_config()

    cmd = ["docker", "run", "--rm", "--network=none"]

    # Add mounts
    for host_path, container_path, mode in mounts:
        cmd.extend(["-v", f"{host_path}:{container_path}:{mode}"])

    # Add environment variables
    cmd.extend(["-e", f"NANOGRIDBOT_IS_MAIN={str(input_data.get('isMain', False)).lower()}"])
    cmd.extend(["-e", f"NANOGRIDBOT_GROUP={input_data.get('groupFolder', '')}"])

    # Add custom environment variables
    if env:
        for key, value in env.items():
            cmd.extend(["-e", f"{key}={value}"])

    # Set timeout
    cmd.extend(["--stop-timeout", str(timeout)])

    # Memory limit
    cmd.extend(["--memory", "2g"])

    # CPU limit
    cmd.extend(["--cpus", "1.0"])

    # Use image
    cmd.append(
        config.container_image if hasattr(config, "container_image") else "nanogridbot-agent:latest"
    )

    return cmd


async def check_docker_available() -> bool:
    """Check if Docker is available.

    Returns:
        True if Docker is available, False otherwise
    """
    try:
        process = await asyncio.create_subprocess_exec(
            "docker",
            "version",
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        await process.communicate()
        return process.returncode == 0
    except FileNotFoundError:
        return False


async def get_container_status(container_name: str) -> Literal["running", "exited", "not_found"]:
    """Get container status.

    Args:
        container_name: Name of container

    Returns:
        Container status
    """
    try:
        process = await asyncio.create_subprocess_exec(
            "docker",
            "inspect",
            "--format={{.State.Status}}",
            container_name,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, _ = await process.communicate()
        if process.returncode == 0:
            status = stdout.decode().strip()
            return status  # type: ignore
        return "not_found"
    except FileNotFoundError:
        return "not_found"


async def cleanup_container(container_name: str) -> None:
    """Clean up a container.

    Args:
        container_name: Name of container to remove
    """
    from loguru import logger

    try:
        process = await asyncio.create_subprocess_exec(
            "docker",
            "rm",
            "-f",
            container_name,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        await process.communicate()
        if process.returncode == 0:
            logger.info(f"Cleaned up container: {container_name}")
    except Exception as e:
        logger.warning(f"Failed to cleanup container {container_name}: {e}")

File: crates/ngb-core/src/container_runner.rs
use std::collections::HashMap;
use std::time::Instant;

use ngb_config::Config;
use ngb_db::{Database, MetricsRepository};
use ngb_types::{ContainerOutput, NanoGridBotError, Result};
use serde_json::json;
use tokio::process::Command;
use tracing::{debug, error, info, warn};

use crate::mount_security::validate_group_mounts;

/// Marker written by the agent container to delimit its JSON output.
pub const OUTPUT_START_MARKER: &str = "---NGB_OUTPUT_START---";
/// End marker.
pub const OUTPUT_END_MARKER: &str = "---NGB_OUTPUT_END---";

/// Run an agent container and return its output.
///
/// Flow: validate mounts → merge env → build docker command →
/// execute with timeout → parse output → record metrics.
#[allow(clippy::too_many_arguments)]
pub async fn run_container_agent(
    group_folder: &str,
    prompt: &str,
    session_id: &str,
    chat_jid: &str,
    is_main: bool,
    additional_mounts: &[HashMap<String, serde_json::Value>],
    timeout_secs: Option<u64>,
    env: &HashMap<String, String>,
    config: &Config,
    db: &Database,
) -> Result<ContainerOutput> {
    let metrics = MetricsRepository::new(db);
    let channel = chat_jid.split(':').next().unwrap_or("unknown");
    let metric_id = metrics
        .record_container_start(group_folder, channel)
        .await?;
    let start = Instant::now();

    let result = run_container_inner(
        group_folder,
        prompt,
        session_id,
        chat_jid,
        is_main,
        additional_mounts,
        timeout_secs,
        env,
        config,
    )
    .await;

    let duration = start.elapsed().as_secs_f64();

    match &result {
        Ok(output) => {
            info!(
                group_folder,
                status = %output.status,
                duration_secs = duration,
                "Container execution completed"
            );
            metrics
                .record_container_end(metric_id, &output.status, Some(duration), None, None, None)
                .await?;
        }
        Err(e) => {
            error!(group_folder, error = %e, "Container execution failed");
            let status = if matches!(e, NanoGridBotError::Timeout(_)) {
                "timeout"
            } else {
                "error"
            };
            metrics
                .record_container_end(
                    metric_id,
                    status,
                    Some(duration),
                    None,
                    None,
                    Some(&e.to_string()),
                )
                .await?;
        }
    }

    result
}

/// Inner function that handles the actual container invocation.
#[allow(clippy::too_many_arguments)]
async fn run_container_inner(
    group_folder: &str,
    prompt: &str,
    session_id: &str,
    chat_jid: &str,
    is_main: bool,
    additional_mounts: &[HashMap<String, serde_json::Value>],
    timeout_secs: Option<u64>,
    env: &HashMap<String, String>,
    config: &Config,
) -> Result<ContainerOutput> {
    // Validate mounts
    let mounts = validate_group_mounts(group_folder, chat_jid, is_main, additional_mounts, config)?;

    // Build the docker command
    let args = build_docker_args(group_folder, &mounts, env, config);

    // Build input JSON
    let input = json!({
        "prompt": prompt,
        "sessionId": session_id,
        "groupFolder": group_folder,
        "chatJid": chat_jid,
        "isMain": is_main,
    });
    let input_bytes = serde_json::to_vec(&input)?;

    debug!(group_folder, args = ?args, "Launching container");

    let timeout = std::time::Duration::from_secs(timeout_secs.unwrap_or(config.container_timeout));

    // Spawn the process
    let mut child = Command::new("docker")
        .args(&args)
        .stdin(std::process::Stdio::piped())
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .spawn()
        .map_err(|e| NanoGridBotError::Container(format!("Failed to spawn docker: {e}")))?;

    // Write input to stdin
    if let Some(mut stdin) = child.stdin.take() {
        use tokio::io::AsyncWriteExt;
        stdin.write_all(&input_bytes).await.map_err(|e| {
            NanoGridBotError::Container(format!("Failed to write to container stdin: {e}"))
        })?;
        drop(stdin);
    }

    // Wait with timeout
    let output = tokio::time::timeout(timeout, child.wait_with_output())
        .await
        .map_err(|_| {
            NanoGridBotError::Timeout(format!(
                "Container timed out after {}s for group {group_folder}",
                timeout.as_secs()
            ))
        })?
        .map_err(|e| NanoGridBotError::Container(format!("Container process error: {e}")))?;

    let stdout = String::from_utf8_lossy(&output.stdout);
    let stderr = String::from_utf8_lossy(&output.stderr);

    if !output.status.success() {
        warn!(
            group_folder,
            exit_code = ?output.status.code(),
            stderr = %stderr,
            "Container exited with non-zero status"
        );
    }

    // Parse output between markers
    parse_container_output(&stdout, &stderr)
}

/// Build docker run arguments (without the "docker" binary itself).
pub fn build_docker_args(
    group_folder: &str,
    mounts: &[crate::mount_security::MountSpec],
    env: &HashMap<String, String>,
    config: &Config,
) -> Vec<String> {
    let container_name = format!("ngb-{}-{}", group_folder, uuid::Uuid::new_v4());
    let mut args = vec![
        "run".to_string(),
        "--rm".to_string(),
        "--name".to_string(),
        container_name,
        "--network=none".to_string(),
        "--memory=2g".to_string(),
        "--cpus=1.0".to_string(),
        "-i".to_string(), // allow stdin
    ];

    // Volume mounts
    for mount in mounts {
        args.push("-v".to_string());
        args.push(mount.to_docker_arg());
    }

    // Environment variables
    for (k, v) in env {
        args.push("-e".to_string());
        args.push(format!("{k}={v}"));
    }

    // Image
    args.push(config.container_image.clone());

    args
}

/// Parse the container stdout for marker-delimited JSON output.
pub fn parse_container_output(stdout: &str, stderr: &str) -> Result<ContainerOutput> {
    // Look for output between markers
    if let Some(start) = stdout.find(OUTPUT_START_MARKER) {
        let after_marker = &stdout[start + OUTPUT_START_MARKER.len()..];
        if let Some(end) = after_marker.find(OUTPUT_END_MARKER) {
            let json_str = after_marker[..end].trim();
            match serde_json::from_str::<ContainerOutput>(json_str) {
                Ok(output) => return Ok(output),
                Err(e) => {
                    warn!(error = %e, "Failed to parse container JSON output");
                }
            }
        }
    }

    // Fallback: try parsing stdout as JSON directly
    if let Ok(output) = serde_json::from_str::<ContainerOutput>(stdout.trim()) {
        return Ok(output);
    }

    // If there's any stdout, treat it as the result
    let trimmed = stdout.trim();
    if !trimmed.is_empty() {
        return Ok(ContainerOutput {
            status: "success".to_string(),
            result: Some(trimmed.to_string()),
            error: None,
            new_session_id: None,
        });
    }

    // No useful output
    let error_msg = if stderr.trim().is_empty() {
        "Container produced no output".to_string()
    } else {
        stderr.trim().to_string()
    };

    Ok(ContainerOutput {
        status: "error".to_string(),
        result: None,
        error: Some(error_msg),
        new_session_id: None,
    })
}

/// Check if Docker is available on the system.
pub async fn check_docker_available() -> Result<bool> {
    match Command::new("docker")
        .arg("version")
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .status()
        .await
    {
        Ok(status) => Ok(status.success()),
        Err(_) => Ok(false),
    }
}

/// Get the status of a named container.
pub async fn get_container_status(name: &str) -> Result<String> {
    let output = Command::new("docker")
        .args(["inspect", "--format", "{{.State.Status}}", name])
        .output()
        .await
        .map_err(|e| NanoGridBotError::Container(format!("Failed to inspect container: {e}")))?;

    if output.status.success() {
        Ok(String::from_utf8_lossy(&output.stdout).trim().to_string())
    } else {
        Ok("not_found".to_string())
    }
}

/// Forcibly remove a container by name.
pub async fn cleanup_container(name: &str) -> Result<()> {
    let output = Command::new("docker")
        .args(["rm", "-f", name])
        .output()
        .await
        .map_err(|e| NanoGridBotError::Container(format!("Failed to cleanup container: {e}")))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        // Ignore "no such container" errors
        if !stderr.contains("No such container") {
            warn!(name, stderr = %stderr, "Container cleanup warning");
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_output_with_markers() {
        let stdout = format!(
            "some log line\n{}\n{{\"status\":\"success\",\"result\":\"Hello\"}}\n{}\ntrailing",
            OUTPUT_START_MARKER, OUTPUT_END_MARKER
        );
        let output = parse_container_output(&stdout, "").unwrap();
        assert_eq!(output.status, "success");
        assert_eq!(output.result, Some("Hello".to_string()));
    }

    #[test]
    fn parse_output_plain_json() {
        let stdout = r#"{"status":"success","result":"Done"}"#;
        let output = parse_container_output(stdout, "").unwrap();
        assert_eq!(output.status, "success");
        assert_eq!(output.result, Some("Done".to_string()));
    }

    #[test]
    fn parse_output_plain_text() {
        let stdout = "Hello, world!";
        let output = parse_container_output(stdout, "").unwrap();
        assert_eq!(output.status, "success");
        assert_eq!(output.result, Some("Hello, world!".to_string()));
    }

    #[test]
    fn parse_output_empty_with_stderr() {
        let output = parse_container_output("", "something went wrong").unwrap();
        assert_eq!(output.status, "error");
        assert_eq!(output.error, Some("something went wrong".to_string()));
    }

    #[test]
    fn parse_output_empty_no_stderr() {
        let output = parse_container_output("", "").unwrap();
        assert_eq!(output.status, "error");
        assert!(output.error.unwrap().contains("no output"));
    }

    #[test]
    fn parse_output_with_new_session_id() {
        let stdout = format!(
            "{}\n{{\"status\":\"success\",\"result\":\"ok\",\"new_session_id\":\"sess-42\"}}\n{}",
            OUTPUT_START_MARKER, OUTPUT_END_MARKER
        );
        let output = parse_container_output(&stdout, "").unwrap();
        assert_eq!(output.new_session_id, Some("sess-42".to_string()));
    }

    #[test]
    fn build_docker_args_basic() {
        use crate::mount_security::{MountMode, MountSpec};
        use std::path::PathBuf;

        let mounts = vec![MountSpec {
            host_path: PathBuf::from("/host/data"),
            container_path: "/workspace/data".to_string(),
            mode: MountMode::ReadOnly,
        }];

        let mut env_map = HashMap::new();
        env_map.insert("API_KEY".to_string(), "secret".to_string());

        let cfg = test_config();
        let args = build_docker_args("test_group", &mounts, &env_map, &cfg);

        assert!(args.contains(&"run".to_string()));
        assert!(args.contains(&"--rm".to_string()));
        assert!(args.contains(&"--network=none".to_string()));
        assert!(args.contains(&"--memory=2g".to_string()));
        assert!(args.contains(&"-v".to_string()));
        assert!(args.contains(&"/host/data:/workspace/data:ro".to_string()));
        assert!(args.contains(&"-e".to_string()));
        assert!(args.contains(&"API_KEY=secret".to_string()));
        assert!(args.contains(&"nanogridbot-agent:latest".to_string()));
    }

    #[test]
    fn build_docker_args_no_env() {
        let cfg = test_config();
        let args = build_docker_args("grp", &[], &HashMap::new(), &cfg);
        // Should not contain -e flag when no env vars
        let e_positions: Vec<_> = args
            .iter()
            .enumerate()
            .filter(|(_, a)| *a == "-e")
            .collect();
        assert!(e_positions.is_empty());
    }

    #[test]
    fn build_docker_args_container_name_format() {
        let cfg = test_config();
        let args = build_docker_args("my_group", &[], &HashMap::new(), &cfg);
        let name_idx = args.iter().position(|a| a == "--name").unwrap();
        let name = &args[name_idx + 1];
        assert!(name.starts_with("ngb-my_group-"));
    }

    #[test]
    fn markers_are_distinct() {
        assert_ne!(OUTPUT_START_MARKER, OUTPUT_END_MARKER);
        assert!(!OUTPUT_START_MARKER.is_empty());
        assert!(!OUTPUT_END_MARKER.is_empty());
    }

    fn test_config() -> Config {
        let base = std::path::PathBuf::from("/tmp/ngb-test");
        Config {
            project_name: "test".to_string(),
            version: "0.0.1".to_string(),
            debug: false,
            base_dir: base.clone(),
            data_dir: base.join("data"),
            store_dir: base.join("store"),
            groups_dir: base.join("groups"),
            db_path: base.join("store/messages.db"),
            whatsapp_session_path: base.join("store/whatsapp_session"),
            openai_api_key: None,
            anthropic_api_key: None,
            telegram_bot_token: None,
            slack_bot_token: None,
            slack_signing_secret: None,
            discord_bot_token: None,
            qq_host: "127.0.0.1".to_string(),
            qq_port: 20000,
            feishu_app_id: None,
            feishu_app_secret: None,
            wecom_corp_id: None,
            wecom_agent_id: None,
            wecom_secret: None,
            dingtalk_app_key: None,
            dingtalk_app_secret: None,
            claude_api_url: "https://api.anthropic.com".to_string(),
            claude_api_version: "2023-06-01".to_string(),
            claude_model: "claude-sonnet-4-20250514".to_string(),
            claude_max_tokens: 4096,
            cli_default_group: "cli".to_string(),
            container_timeout: 300,
            container_max_output_size: 100_000,
            container_max_concurrent: 5,
            container_image: "nanogridbot-agent:latest".to_string(),
            assistant_name: "Andy".to_string(),
            trigger_pattern: None,
            poll_interval: 2000,
            max_messages_per_minute: 10,
            message_cache_size: 1000,
            batch_size: 100,
            db_connection_pool_size: 5,
            ipc_file_buffer_size: 8192,
            log_level: "INFO".to_string(),
            log_format: "default".to_string(),
            log_rotation: "10 MB".to_string(),
            log_retention: "7 days".to_string(),
            web_host: "0.0.0.0".to_string(),
            web_port: 8080,
        }
    }
}
